# Table of Contents
1. Import libraries
2. Import data
3. Exercise
4. Export

# 1.0 Import libraries

In [3]:
# Importing libraries
import pandas as pd
import numpy as np
import os

# 2.0 Import data

In [5]:
# Define project folder path
path = r'/Users/sharnti/Desktop/CareerFoundry/Data Immersion/Achievement 4/Instacart Basket Analysis'

In [6]:
# Import orders_products_combined dataframe
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge_derived.pkl'))

In [7]:
# View rows, columns
ords_prods_merge.shape

(32404859, 20)

In [8]:
# View sample of import to check it worked
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                     NaN         True         196                  1   
1                     NaN         True       14084                  2   
2                     NaN         True       12427                  3   
3                     NaN         True       26088                  4   
4                     NaN         True       26405                  5   

   reordered _merge                             product_name  aisle_id  \
0          0   both                                     Soda        77   
1          0   both  Organic Unsweetened Vanilla Almond Milk        91   
2          0   both                      Original Beef Jerky        23   
3          0   both               Aged White Cheddar Popcorn        23   
4          0   both         XL Pick-A-Size Paper Towel Rolls        54   

   department_id  prices match    price_range_loc     busiest_day  \
0              7     9.0  both  Mid-range product  Regularly busy   
1             16    12.5  both  Mid-range product  Regularly busy   
2             19     4.4  both  Low-range product  Regularly busy   
3             19     4.7  both  Low-range product  Regularly busy   
4             17     1.0  both  Low-range product  Regularly busy   

          busiest_days busiest_period_of_day  
0  Regularly busy days        Average orders  
1  Regularly busy days        Average orders  
2  Regularly busy days        Average orders  
3  Regularly busy days        Average orders  
4  Regularly busy days        Average orders

# 3.0 Exercise

## 3.1 Create dataframe subset

In [11]:
# Create subset of dataframe with only 1 million rows.
df = ords_prods_merge[:1000000]

In [12]:
# Check shape of dataframe
df.shape

(1000000, 20)

In [13]:
# Check column-names and rows
df.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                     NaN         True         196                  1   
1                     NaN         True       14084                  2   
2                     NaN         True       12427                  3   
3                     NaN         True       26088                  4   
4                     NaN         True       26405                  5   

   reordered _merge                             product_name  aisle_id  \
0          0   both                                     Soda        77   
1          0   both  Organic Unsweetened Vanilla Almond Milk        91   
2          0   both                      Original Beef Jerky        23   
3          0   both               Aged White Cheddar Popcorn        23   
4          0   both         XL Pick-A-Size Paper Towel Rolls        54   

   department_id  prices match    price_range_loc     busiest_day  \
0              7     9.0  both  Mid-range product  Regularly busy   
1             16    12.5  both  Mid-range product  Regularly busy   
2             19     4.4  both  Low-range product  Regularly busy   
3             19     4.7  both  Low-range product  Regularly busy   
4             17     1.0  both  Low-range product  Regularly busy   

          busiest_days busiest_period_of_day  
0  Regularly busy days        Average orders  
1  Regularly busy days        Average orders  
2  Regularly busy days        Average orders  
3  Regularly busy days        Average orders  
4  Regularly busy days        Average orders

## 3.2 Grouping Data with pandas

In [15]:
# Group the dataframe by the column 'product_name'
df.groupby('product_name')

The __groupby__ function doesn't do much on it's own. We can see that it's created an object, but you need to use it with another function to see the results of your __groupby__.

## 3.3 Aggregating Data with agg()

Aggregation functions create __summaries__ of __selected columns__ and __store these summary values__ in __new columns__ (i.e. mean, median, maximum, or minimum).

Aggregations are important not only for descriptive purposes, but also as a standardization for data to be used for predictive modelling (as it reduces the amount of random variation in the data). 

### 3.3.1 Performing single aggregations

In [20]:
# Split data into groups based on department_id, then calculate the mean of the order_number column for each of these groups.
df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                14.800024
2                17.091743
3                17.913544
4                17.893092
5                15.214270
6                15.382135
7                17.694027
8                16.458105
9                15.957363
10               20.091818
11               16.482026
12               15.615061
13               16.484023
14               17.524632
15               15.691875
16               18.014071
17               16.150593
18               19.602850
19               17.631340
20               17.138607
21               21.956893

In [21]:
# Same calculation as above using mean() instead of agg()
df.groupby('department_id')['order_number'].mean()

department_id
1     14.800024
2     17.091743
3     17.913544
4     17.893092
5     15.214270
6     15.382135
7     17.694027
8     16.458105
9     15.957363
10    20.091818
11    16.482026
12    15.615061
13    16.484023
14    17.524632
15    15.691875
16    18.014071
17    16.150593
18    19.602850
19    17.631340
20    17.138607
21    21.956893
Name: order_number, dtype: float64

In [22]:
# This can also be written with dot rather than square notation.
df.groupby('department_id').order_number.mean()

department_id
1     14.800024
2     17.091743
3     17.913544
4     17.893092
5     15.214270
6     15.382135
7     17.694027
8     16.458105
9     15.957363
10    20.091818
11    16.482026
12    15.615061
13    16.484023
14    17.524632
15    15.691875
16    18.014071
17    16.150593
18    19.602850
19    17.631340
20    17.138607
21    21.956893
Name: order_number, dtype: float64

__NB:__ Square bracket are recommended because:
- Stand out, so easy for others to read in your code
- The default role of [] square brackets is for indexing. Dots have various roles

### 3.3.2 Performing multiple aggregations

In [25]:
# Performing multiple aggregations at once
df.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

order_number        
                      mean min max
department_id                     
1                14.800024   1  99
2                17.091743   1  98
3                17.913544   1  99
4                17.893092   1  99
5                15.214270   1  99
6                15.382135   1  99
7                17.694027   1  99
8                16.458105   1  91
9                15.957363   1  99
10               20.091818   1  99
11               16.482026   1  99
12               15.615061   1  99
13               16.484023   1  99
14               17.524632   1  99
15               15.691875   1  99
16               18.014071   1  99
17               16.150593   1  99
18               19.602850   1  99
19               17.631340   1  99
20               17.138607   1  99
21               21.956893   1  97

## 3.4 Aggregating Data with transform()

Here is a briefy explanation fo the code below:
1. A new column called “max_order” is created, which will be what stores the maximum order number for each user.
2. Then, the ords_prods_merge dataframe is grouped by the “user_id” column.
3. And finally, the transform() function is applied on the “order_number” column with the np.max argument.

In [28]:
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform('max')

In [29]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                     NaN         True         196                  1   
1                     NaN         True       14084                  2   
2                     NaN         True       12427                  3   
3                     NaN         True       26088                  4   
4                     NaN         True       26405                  5   

   reordered  ...                             product_name aisle_id  \
0          0  ...                                     Soda       77   
1          0  ...  Organic Unsweetened Vanilla Almond Milk       91   
2          0  ...                      Original Beef Jerky       23   
3          0  ...               Aged White Cheddar Popcorn       23   
4          0  ...         XL Pick-A-Size Paper Towel Rolls       54   

   department_id  prices  match    price_range_loc     busiest_day  \
0              7     9.0   both  Mid-range product  Regularly busy   
1             16    12.5   both  Mid-range product  Regularly busy   
2             19     4.4   both  Low-range product  Regularly busy   
3             19     4.7   both  Low-range product  Regularly busy   
4             17     1.0   both  Low-range product  Regularly busy   

          busiest_days busiest_period_of_day max_order  
0  Regularly busy days        Average orders        10  
1  Regularly busy days        Average orders        10  
2  Regularly busy days        Average orders        10  
3  Regularly busy days        Average orders        10  
4  Regularly busy days        Average orders        10  

[5 rows x 21 columns]

In [30]:
ords_prods_merge.head(100)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2539329        1             1                   2                  8   
2    2539329        1             1                   2                  8   
3    2539329        1             1                   2                  8   
4    2539329        1             1                   2                  8   
..       ...      ...           ...                 ...                ...   
95    738281        2             4                   2                 10   
96   1673511        2             5                   3                 11   
97   1673511        2             5                   3                 11   
98   1673511        2             5                   3                 11   
99   1673511        2             5                   3                 11   

    days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                      NaN         True         196                  1   
1                      NaN         True       14084                  2   
2                      NaN         True       12427                  3   
3                      NaN         True       26088                  4   
4                      NaN         True       26405                  5   
..                     ...          ...         ...                ...   
95                     8.0        False       21150                 13   
96                     8.0        False       47144                  1   
97                     8.0        False        5322                  2   
98                     8.0        False       17224                  3   
99                     8.0        False       38656                  4   

    reordered  ...                                       product_name  \
0           0  ...                                               Soda   
1           0  ...            Organic Unsweetened Vanilla Almond Milk   
2           0  ...                                Original Beef Jerky   
3           0  ...                         Aged White Cheddar Popcorn   
4           0  ...                   XL Pick-A-Size Paper Towel Rolls   
..        ...  ...                                                ...   
95          0  ...                            Fire Grilled Steak Bowl   
96          0  ...     Unsweetened Original Almond Breeze Almond Milk   
97          0  ...  Gluten Free Dark Chocolate Chunk Chewy with a ...   
98          0  ...                   Oats & Honey Gluten Free Granola   
99          0  ...  Gluten Free Peanut Butter Dark Chocolate Chewy...   

   aisle_id  department_id  prices  match    price_range_loc     busiest_day  \
0        77              7     9.0   both  Mid-range product  Regularly busy   
1        91             16    12.5   both  Mid-range product  Regularly busy   
2        23             19     4.4   both  Low-range product  Regularly busy   
3        23             19     4.7   both  Low-range product  Regularly busy   
4        54             17     1.0   both  Low-range product  Regularly busy   
..      ...            ...     ...    ...                ...             ...   
95       38              1     5.9   both  Mid-range product  Regularly busy   
96       91             16    14.0   both  Mid-range product  Regularly busy   
97        3             19     2.9   both  Low-range product  Regularly busy   
98        3             19     1.6   both  Low-range product  Regularly busy   
99        3             19     3.4   both  Low-range product  Regularly busy   

           busiest_days busiest_period_of_day max_order  
0   Regularly busy days        Average orders        10  
1   Regularly busy days        Average orders        10  
2   Regularly busy days        Average orders        10  
3   Regularly busy days        Average orders        10  
4   Regularly busy days        Average orders     

In [31]:
pd.options.display.max_rows = None

## 3.5 Deriving Columns with loc()

In [33]:
# Create a flag that assigns a “loyalty” label to a user ID based on its corresponding max order value (from 3.4).
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [34]:
# Check count of each flag
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

In [35]:
# Return only the top 60 customers from the dataframe for the columns 'user_id', 'loyalty_flag' and 'order_number')
ords_prods_merge[['user_id', 'loyalty_flag', 'order_number']].head(60)

user_id      loyalty_flag  order_number
0         1      New customer             1
1         1      New customer             1
2         1      New customer             1
3         1      New customer             1
4         1      New customer             1
5         1      New customer             2
6         1      New customer             2
7         1      New customer             2
8         1      New customer             2
9         1      New customer             2
10        1      New customer             2
11        1      New customer             3
12        1      New customer             3
13        1      New customer             3
14        1      New customer             3
15        1      New customer             3
16        1      New customer             4
17        1      New customer             4
18        1      New customer             4
19        1      New customer             4
20        1      New customer             4
21        1      New customer             5
22        1      New customer             5
23        1      New customer             5
24        1      New customer             5
25        1      New customer             5
26        1      New customer             5
27        1      New customer             5
28        1      New customer             5
29        1      New customer             6
30        1      New customer             6
31        1      New customer             6
32        1      New customer             6
33        1      New customer             7
34        1      New customer             7
35        1      New customer             7
36        1      New customer             7
37        1      New customer             7
38        1      New customer             8
39        1      New customer             8
40        1      New customer             8
41        1      New customer             8
42        1      New customer             8
43        1      New customer             8
44        1      New customer             9
45        1      New customer             9
46        1      New customer             9
47        1      New customer             9
48        1      New customer             9
49        1      New customer             9
50        1      New customer            10
51        1      New customer            10
52        1      New customer            10
53        1      New customer            10
54        1      New customer            10
55        1      New customer            10
56        1      New customer            10
57        1      New customer            10
58        1      New customer            10
59        2  Regular customer             1

# 4.0 Task

## Question 1

Libraries import and data import can be viewed above.

## Question 2 & 3

Here are the results from the df (sample of ords_prods_merge):

In [41]:
# Split data into groups based on department_id, then calculate the mean of the order_number column for each of these groups.
df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                14.800024
2                17.091743
3                17.913544
4                17.893092
5                15.214270
6                15.382135
7                17.694027
8                16.458105
9                15.957363
10               20.091818
11               16.482026
12               15.615061
13               16.484023
14               17.524632
15               15.691875
16               18.014071
17               16.150593
18               19.602850
19               17.631340
20               17.138607
21               21.956893

Here are the results when applying this to the whole dataframe (ords_prods_merge):

In [43]:
# Split data into groups based on department_id, then calculate the mean of the order_number column for each of these groups.
ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

The results from the subset and the entire dataframe are quite similar. From most departments, they are within 1.0 unit difference.

## Question 4

### Create a loyalty flag for existing customers

Please refer back to steps 3.4 and 3.5 to view code used to create the loyalty flag.

## Question 5

The marketing team at Instacart wants to know whether there’s a difference between the spending habits of the three types of customers you identified. Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). What you’re trying to determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [50]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  first_order  product_id  add_to_cart_order  \
0                     NaN         True         196                  1   
1                     NaN         True       14084                  2   
2                     NaN         True       12427                  3   
3                     NaN         True       26088                  4   
4                     NaN         True       26405                  5   

   reordered  ... aisle_id department_id  prices  match    price_range_loc  \
0          0  ...       77             7     9.0   both  Mid-range product   
1          0  ...       91            16    12.5   both  Mid-range product   
2          0  ...       23            19     4.4   both  Low-range product   
3          0  ...       23            19     4.7   both  Low-range product   
4          0  ...       54            17     1.0   both  Low-range product   

      busiest_day         busiest_days busiest_period_of_day max_order  \
0  Regularly busy  Regularly busy days        Average orders        10   
1  Regularly busy  Regularly busy days        Average orders        10   
2  Regularly busy  Regularly busy days        Average orders        10   
3  Regularly busy  Regularly busy days        Average orders        10   
4  Regularly busy  Regularly busy days        Average orders        10   

   loyalty_flag  
0  New customer  
1  New customer  
2  New customer  
3  New customer  
4  New customer  

[5 rows x 22 columns]

In [51]:
ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

I wouldn't say there is a sizeable difference in spending habits of customers based on their loyalty, as there is only approximately 3 dollars difference between their averages. More worrying is the max price listed for all three loyalty groups is __$99,999__ - this seems _way_ too high for grocery items and is likely an error in the data.

## Question 6

### Create a spending flag based on average prices

Group users based on the average price across all their orders using the following criteria:
- If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”
- If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”

In [56]:
# Create column with calculation of average price by a user.
ords_prods_merge['avg_price'] = ords_prods_merge.groupby(['user_id'])['prices'].transform('mean')

In [57]:
# Check that the 'avg_price' column has been added to the dataframe and that the calculation makes sense.
ords_prods_merge[['user_id','prices','avg_price']].head(60)

user_id  prices  avg_price
0         1     9.0   6.367797
1         1    12.5   6.367797
2         1     4.4   6.367797
3         1     4.7   6.367797
4         1     1.0   6.367797
5         1     9.0   6.367797
6         1     3.0   6.367797
7         1     4.4   6.367797
8         1    10.3   6.367797
9         1     4.7   6.367797
10        1     4.0   6.367797
11        1     9.0   6.367797
12        1     4.4   6.367797
13        1     3.0   6.367797
14        1     8.6   6.367797
15        1     3.2   6.367797
16        1     9.0   6.367797
17        1     4.4   6.367797
18        1     3.0   6.367797
19        1     8.6   6.367797
20        1     1.0   6.367797
21        1     9.0   6.367797
22        1     4.4   6.367797
23        1     3.0   6.367797
24        1     8.6   6.367797
25        1     2.7   6.367797
26        1    14.0   6.367797
27        1     5.0   6.367797
28        1    10.3   6.367797
29        1     9.0   6.367797
30        1     4.4   6.367797
31        1     3.0   6.367797
32        1     8.6   6.367797
33        1     9.0   6.367797
34        1     3.0   6.367797
35        1     4.4   6.367797
36        1     8.6   6.367797
37        1     4.0   6.367797
38        1     4.4   6.367797
39        1     9.0   6.367797
40        1     3.0   6.367797
41        1     8.6   6.367797
42        1    13.4   6.367797
43        1     1.8   6.367797
44        1     1.8   6.367797
45        1    13.4   6.367797
46        1     8.6   6.367797
47        1     9.0   6.367797
48        1     3.0   6.367797
49        1     4.4   6.367797
50        1     9.0   6.367797
51        1    13.4   6.367797
52        1     6.8   6.367797
53        1    12.6   6.367797
54        1     8.6   6.367797
55        1     3.0   6.367797
56        1     1.3   6.367797
57        1     4.0   6.367797
58        1     4.4   6.367797
59        2     5.2   7.515897

Whilst I haven't manually double-checked the results of the avg_price column for user 1, from scrolling through the prices it seems about right - there are a lot of values around 4 with some larger values pulling the average up to 6.37.

In [59]:
# Create a flag that assigns a “spender” label to a user ID based on its corresponding average price value.
ords_prods_merge.loc[ords_prods_merge['avg_price'] < 10, 'spender_flag'] = 'Low spender'
ords_prods_merge.loc[ords_prods_merge['avg_price'] >= 10, 'spender_flag'] = 'High customer'

In [60]:
# Check count of each flag
ords_prods_merge['spender_flag'].value_counts(dropna = False)

spender_flag
Low spender      31770614
High customer      634245
Name: count, dtype: int64

In [61]:
# Check flag is working
ords_prods_merge[['user_id','spender_flag']].head(20)

user_id spender_flag
0         1  Low spender
1         1  Low spender
2         1  Low spender
3         1  Low spender
4         1  Low spender
5         1  Low spender
6         1  Low spender
7         1  Low spender
8         1  Low spender
9         1  Low spender
10        1  Low spender
11        1  Low spender
12        1  Low spender
13        1  Low spender
14        1  Low spender
15        1  Low spender
16        1  Low spender
17        1  Low spender
18        1  Low spender
19        1  Low spender

## Question 7

### Create an order frequency flag based on the median of days_since_prior_order

Group users based on the median of days_since_prior_order using the following criteria:
- If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”
- If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”
- If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”

In [65]:
# Create column with calculation of average price by a user.
ords_prods_merge['median_days_since_prior_order'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform('median')

In [66]:
# Check that the 'avg_price' column has been added to the dataframe and that the calculation makes sense.
ords_prods_merge[['user_id','days_since_prior_order','median_days_since_prior_order']].head(60)

user_id  days_since_prior_order  median_days_since_prior_order
0         1                     NaN                           20.5
1         1                     NaN                           20.5
2         1                     NaN                           20.5
3         1                     NaN                           20.5
4         1                     NaN                           20.5
5         1                    15.0                           20.5
6         1                    15.0                           20.5
7         1                    15.0                           20.5
8         1                    15.0                           20.5
9         1                    15.0                           20.5
10        1                    15.0                           20.5
11        1                    21.0                           20.5
12        1                    21.0                           20.5
13        1                    21.0                           20.5
14        1                    21.0                           20.5
15        1                    21.0                           20.5
16        1                    29.0                           20.5
17        1                    29.0                           20.5
18        1                    29.0                           20.5
19        1                    29.0                           20.5
20        1                    29.0                           20.5
21        1                    28.0                           20.5
22        1                    28.0                           20.5
23        1                    28.0                           20.5
24        1                    28.0                           20.5
25        1                    28.0                           20.5
26        1                    28.0                           20.5
27        1                    28.0                           20.5
28        1                    28.0                           20.5
29        1                    19.0                           20.5
30        1                    19.0                           20.5
31        1                    19.0                           20.5
32        1                    19.0                           20.5
33        1                    20.0                           20.5
34        1                    20.0                           20.5
35        1                    20.0                           20.5
36        1                    20.0                           20.5
37        1                    20.0                           20.5
38        1                    14.0                           20.5
39        1                    14.0                           20.5
40        1                    14.0                           20.5
41        1                    14.0                           20.5
42        1                    14.0                           20.5
43        1                    14.0                           20.5
44        1                     0.0                           20.5
45        1                     0.0                           20.5
46        1                     0.0                           20.5
47        1                     0.0                           20.5
48        1                     0.0                           20.5
49        1                     0.0                           20.5
50        1                    30.0                           20.5
51        1                    30.0                           20.5
52        1                    30.0                           20.5
53        1                    30.0                           20.5
54        1                    30.0                           20.5
55        1                    30.0                           20.5
56        1                    30.0                           20.5
57        1                    30.0                           20.5
58        1                    30.0                

Whilst I haven't manually checked the above, a cursory overview seems to indicate the calculation has worked.

In [68]:
# Create a flag that assigns a “spender” label to a user ID based on its corresponding average price value.
ords_prods_merge.loc[ords_prods_merge['median_days_since_prior_order'] > 20, 'frequency_flag'] = 'Non-frequent customer'
ords_prods_merge.loc[(ords_prods_merge['median_days_since_prior_order'] <= 20) & (ords_prods_merge['median_days_since_prior_order'] > 10), 'frequency_flag'] = 'Regular customer'
ords_prods_merge.loc[ords_prods_merge['median_days_since_prior_order'] <= 10, 'frequency_flag'] = 'Frequent customer'

In [69]:
# Check count of each flag
ords_prods_merge['frequency_flag'].value_counts(dropna = False)

frequency_flag
Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: count, dtype: int64

In [70]:
# Check flag is working
ords_prods_merge[['user_id','frequency_flag']].head(20)

user_id         frequency_flag
0         1  Non-frequent customer
1         1  Non-frequent customer
2         1  Non-frequent customer
3         1  Non-frequent customer
4         1  Non-frequent customer
5         1  Non-frequent customer
6         1  Non-frequent customer
7         1  Non-frequent customer
8         1  Non-frequent customer
9         1  Non-frequent customer
10        1  Non-frequent customer
11        1  Non-frequent customer
12        1  Non-frequent customer
13        1  Non-frequent customer
14        1  Non-frequent customer
15        1  Non-frequent customer
16        1  Non-frequent customer
17        1  Non-frequent customer
18        1  Non-frequent customer
19        1  Non-frequent customer

# 5.0 Assigning NaN to Price Outlier

In [72]:
# Check for price outliers (where price is greater than 100).
ords_prods_merge[['prices']].loc[ords_prods_merge['prices'] > 100]

prices
1576      14900.0
1638      14900.0
16522     14900.0
16528     14900.0
53672     14900.0
91430     14900.0
91460     14900.0
98762     14900.0
98800     14900.0
98980     14900.0
99030     14900.0
122849    14900.0
127212    14900.0
127231    14900.0
127250    14900.0
127262    14900.0
127278    14900.0
127287    14900.0
127313    14900.0
127325    14900.0
127365    14900.0
127388    14900.0
127417    14900.0
128911    99999.0
128924    99999.0
136836    14900.0
136865    14900.0
136885    14900.0
136920    14900.0
137044    14900.0
137118    14900.0
142302    14900.0
156000    14900.0
156038    14900.0
157449    14900.0
172029    14900.0
181571    14900.0
191245    14900.0
191268    14900.0
221613    14900.0
221710    14900.0
250278    14900.0
250330    14900.0
273994    14900.0
282631    14900.0
284218    99999.0
284229    99999.0
284252    99999.0
289857    14900.0
290204    14900.0
307363    14900.0
360926    14900.0
371481    14900.0
409993    14900.0
418613    14900.0
426987    14900.0
459750    14900.0
465704    14900.0
468510    14900.0
480656    14900.0
493387    14900.0
499076    99999.0
499119    99999.0
499141    99999.0
499158    99999.0
499180    99999.0
524757    14900.0
527458    99999.0
553878    14900.0
553907    14900.0
553972    14900.0
554866    14900.0
554871    14900.0
554951    14900.0
568989    14900.0
574165    14900.0
574207    14900.0
574229    14900.0
574266    14900.0
628557    14900.0
662469    14900.0
662507    14900.0
662568    14900.0
662704    14900.0
662883    14900.0
687351    14900.0
691237    14900.0
691300    14900.0
691396    14900.0
691481    14900.0
691540    14900.0
691569    14900.0
763748    14900.0
763852    14900.0
767471    99999.0
767485    99999.0
768338    14900.0
782661    14900.0
782703    14900.0
795100    14900.0
795154    14900.0
795174    14900.0
795193    14900.0
795202    14900.0
795226    14900.0
797951    14900.0
797993    14900.0
816396    14900.0
816399    14900.0
816410    14900.0
816421    14900.0
816462    14900.0
826196    14900.0
837398    14900.0
837419    14900.0
837473    14900.0
837553    14900.0
837645    14900.0
840391    14900.0
840417    14900.0
840439    14900.0
853726    14900.0
853845    14900.0
854472    14900.0
854506    14900.0
854581    14900.0
862433    14900.0
862447    14900.0
903468    14900.0
903477    14900.0
903486    14900.0
903498    14900.0
903520    14900.0
903538    14900.0
903545    14900.0
903563    14900.0
908979    14900.0
909146    14900.0
923722    14900.0
923756    14900.0
923801    14900.0
923807    14900.0
924006    14900.0
925487    14900.0
925519    14900.0
930698    14900.0
930703    14900.0
930845    14900.0
942512    14900.0
953246    99999.0
958929    14900.0
958969    14900.0
958974    14900.0
959020    14900.0
960274    14900.0
960348    14900.0
960509    99999.0
1027586   14900.0
1027590   14900.0
1043035   14900.0
1059806   99999.0
1080656   14900.0
1080724   14900.0
1086467   99999.0
1099346   14900.0
1102681   14900.0
1102713   14900.0
1102745   14900.0
1102749   14900.0
1102765   14900.0
1102782   14900.0
1102783   14900.0
1102804   14900.0
1102819   14900.0
1102860   14900.0
1102886   14900.0
1115224   14900.0
1115265   14900.0
1115309   14900.0
1115499   14900.0
1115570   14900.0
1115582   14900.0
1115615   14900.0
1115641   14900.0
1115725   14900.0
1115772   14900.0
1118558   14900.0
1118798   14900.0
1118817   14900.0
1122691   14900.0
1122723   14900.0
1128759   14900.0
1128868   14900.0
1174861   99999.0
1193359   14900.0
1198641   14900.0
1198774   14900.0
1198783   14900.0
1198790   14900.0
1198797   14900.0
1203898   14900.0
1209997   14900.0
1225382   14900.0
1225397   14900.0
1258081   14900.0
1262258   14900.0
1287028   14900.0
1287063   14900.0
1295815   99999.0
1295826   99999.0
1297895   99999.0
1297910   99999.0
1297916   99999.0
1298008   99999.0
1298029   99999.0
1298130   14900.0
1327985   14900.0
1346942   14900.0
1349187   14900.0
1352800   14900.0
1368657   14900.0
1368674   14900

In [112]:
# Assign missing values to the above outliers
ords_prods_merge.loc[ords_prods_merge['prices'] >100, 'prices'] = np.nan

In [120]:
# Check this has worked
ords_prods_merge['prices'].max()

25.0

# 6.0 Export dataframe

In [123]:
ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_grouped.pkl'))